### PR TITLE
timetracking: add farming contract functionality

### LINKED ISSUE #7634: Farming contract overlay
My feature suggestion is basically an exact copy of the slayer task overlay, except for the farming contracts, I find myself always forgetting what my current task is, and feel it would be very convenient.
runelite_2019-01-31_00-07-41

### PR DESCRIPTION
**Pull request by melkypie**
Closes #7634

Allotment | Berries | Flower | Herbs | Special | Trees
-- | -- | -- | -- | -- | --
~~Cabbages~~ | ~~Cadava berries~~ | ~~Limpwurt roots~~ | ~~Avantoe~~ | ~~Cactus~~ | ~~Apple tree~~
~~Onions~~ | ~~Dwellberries~~ | ~~Marigolds~~ | ~~Cadantine~~ | ~~Potato cacti~~ | ~~Banana tree~~
~~Potatoes~~ | ~~Jangerberries~~ | ~~Nasturtiums~~ | ~~Dwarf weed~~ |   | ~~Celastrus tree~~
~~Snape grass~~ | ~~Poison ivy berries~~ | ~~Rosemary~~ | ~~Guam~~ |   | ~~Curry tree~~
~~Strawberries~~ | ~~Redberries~~ | ~~White lillies~~ | ~~Harralander~~ |   | ~~Dragonfruit tree~~
~~Sweetcorn~~ | ~~White berries~~ | ~~Woad~~ | ~~Irit~~ |   | ~~Magic tree~~
~~Tomatoes~~ |   |   | ~~Kwuarm~~ |   | ~~Maple tree~~
~~Watermelons~~ |   |   | ~~Lantadyme~~ |   | ~~Oak tree~~
  |   |   | ~~Marrentill~~ |   | ~~Orange tree~~
  |   |   | ~~Ranarr~~ |   | ~~Palm tree~~
  |   |   | ~~Snapdragon~~ |   | ~~Papaya tree~~
  |   |   | ~~Tarromin~~ |   | ~~Pineapple plant~~
  |   |   | ~~Toadflax~~ |   | ~~Redwood tree~~
  |   |   | ~~Torstol~~ |   | ~~Willow tree~~
  |   |   |   |   | ~~Yew tree~~

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+import net.runelite.client.plugins.timetracking.TimeTrackingPlugin;
+import net.runelite.client.ui.overlay.infobox.InfoBoxManager;
+import net.runelite.client.util.Text;
+
+@Slf4j
+public class FarmingContractManager
+{
+	private static final int GUILDMASTER_JANE_WIDGET_MODEL_ID = 8628;
+	private static final int FARMING_GUILD_REGION_ID = 4922;
+	private static final String CONTRACT_ASSIGN_REGEX = "(?:We need you to grow|Please could you grow) (?:some|a|an) ([a-zA-Z ]+)(?: for us\\?|\\.)";
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:56]: Precompile this

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+			infoBox = new FarmingContractInfoBox(itemManager.getImage(contract.getProduce().getItemID()), plugin, contract, config, this);
+			infoBoxManager.addInfoBox(infoBox);
+		}
+	}
+
+	private void handleGuildmasterJaneWidgetDialog()
+	{
+		Widget npcDialog = client.getWidget(WidgetInfo.DIALOG_NPC_HEAD_MODEL);
+
+		if (npcDialog == null || npcDialog.getModelId() != GUILDMASTER_JANE_WIDGET_MODEL_ID)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:195]: We need to grow a NullNpcID instead of having random npc constants in plugins

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+			return;
+		}
+
+		String dialogText = Text.removeTags(client.getWidget(WidgetInfo.DIALOG_NPC_TEXT).getText());
+		log.debug(dialogText + " ");
+
+		if (dialogText.equals(CONTRACT_REWARDED))
+		{
+			setContract(null);
+			;
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:206]: ;

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+	private FarmingContract contract = null;
+	private FarmingContractInfoBox infoBox;
+
+	@Getter
+	private long completionTime;
+
+	@Inject
+	public FarmingContractManager(
+		@Nonnull Client client,
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:83]: `@Nonnull` is unneeded here

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractInfoBox.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import net.runelite.client.util.Text;
+
+class FarmingContractInfoBox extends InfoBox
+{
+	@Getter
+	private final FarmingContract contract;
+	private final FarmingContractManager manager;
+	private final TimeTrackingConfig config;
+
+	FarmingContractInfoBox(@Nonnull BufferedImage image, @Nonnull Plugin plugin, @Nonnull FarmingContract contract, @Nonnull TimeTrackingConfig config, @Nonnull FarmingContractManager manager)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractInfoBox.java:47]: `@Nonnull` is unneeded here

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContract.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+package net.runelite.client.plugins.timetracking.farming;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+enum FarmingContract
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContract.java:35]: Why isn't this just part of Produce?

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/OverviewItemPanel.java [hunk 7]
```diff
@@ -43,11 +43,27 @@
 
 class OverviewItemPanel extends JPanel
 {
+	interface TabSwitch
+	{
+		void selected();
+	}
+
+	interface Selectable
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/OverviewItemPanel.java:51]: its probably more clear to just use a BooleanPredicate than this

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/OverviewItemPanel.java [hunk 8]
```diff
@@ -43,11 +43,27 @@
 
 class OverviewItemPanel extends JPanel
 {
+	interface TabSwitch
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/OverviewItemPanel.java:46]: this can just be a Runnable instead

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/OverviewItemPanel.java [hunk 9]
```diff
@@ -56,16 +72,23 @@
 
 	OverviewItemPanel(ItemManager itemManager, TimeTrackingPanel pluginPanel, Tab tab, String title)
 	{
+		this(itemManager, () -> pluginPanel.switchTab(tab), () -> true, tab.getItemID(), title);
+	}
+
+	OverviewItemPanel(ItemManager itemManager, TabSwitch tabSwitch, Selectable selectable, int iconItemID, String title)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/OverviewItemPanel.java:78]: tabSwitch -> onTabSwitched

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/TimeTrackingPlugin.java [hunk 10]
```diff
@@ -96,6 +102,7 @@
 
 	private WorldPoint lastTickLocation;
 	private boolean lastTickPostLogin;
+	private final String CONTRACT_COMPLETED = "You've completed a Farming Guild Contract. You should return to Guildmaster Jane.";
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/TimeTrackingPlugin.java:105]: static + move to the top of the class

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+@Slf4j
+public class FarmingContractManager
+{
+	private static final int GUILDMASTER_JANE_WIDGET_MODEL_ID = 8628;
+	private static final int FARMING_GUILD_REGION_ID = 4922;
+	private static final String CONTRACT_ASSIGN_REGEX = "(?:We need you to grow|Please could you grow) (?:some|a|an) ([a-zA-Z ]+)(?: for us\\?|\\.)";
+	private static final String CONTRACT_REWARDED = "You'll be wanting a reward then. Here you go.";
+
+	private static final String FARMING_REGION_NAME = "Farming Guild";
+	private static final String HTML_STRIP = "\\<[^>]*>";
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:60]: this is unused

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingTabPanel.java [hunk 12]
```diff
@@ -40,25 +41,29 @@
...
 	private final List<TimeablePanel<FarmingPatch>> patchPanels;
+	private final FarmingContractManager farmingContractManager;
 
 	FarmingTabPanel(
 		FarmingTracker farmingTracker,
 		ItemManager itemManager,
 		TimeTrackingConfig config,
-		Set<FarmingPatch> patches
+		Set<FarmingPatch> patches,
+		FarmingContractManager manager
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingTabPanel.java:59]:
```suggestion
		FarmingContractManager farmingContractManager
```

File: runelite-client/src/main/java/net/runelite/client/util/Text.java [hunk 13]
```diff
@@ -206,4 +210,49 @@ public static boolean matchesSearchTerms(String[] searchTerms, final Collection<
...
 	}
+
+	/**
+	 * Returns a formatted time string
+	 *
+	 * @param remainingSeconds
+	 * @param useRelativeTime
+	 * @return a relative time string (1d 2h 3m) if useRelativeTime is true, date time string (Sunday at 3:05) (at 17:30) if useRelativeTime is false
+	 */
+
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/util/Text.java:221]: remove this newline

File: runelite-client/src/main/java/net/runelite/client/util/Text.java [hunk 14]
```diff
@@ -206,4 +210,49 @@ public static boolean matchesSearchTerms(String[] searchTerms, final Collection<
...
+	public static String getFormattedEstimate(long remainingSeconds, boolean useRelativeTime)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/util/Text.java:222]: why did you move this?

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+			summary = SummaryState.UNKNOWN;
+			return;
+		}
+
+		PatchImplementation patchImplementation = contract.getPatchImplementation();
+		List<PatchPrediction> predictions = new ArrayList<>();
+
+		for (FarmingRegion farmingRegion : farmingWorld.getRegions().values())
+		{
+			if (!farmingRegion.getName().equalsIgnoreCase(FARMING_REGION_NAME))
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:253]: just store this in FarmingWorld and add a getter so you don't have to iterate

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+	public String getContractName()
+	{
+		return hasContract() ? contract.getName() : null;
+	}
+
+	public boolean shouldHighlightFarmingTabPanel(@Nonnull FarmingPatch patch)
+	{
+		PatchPrediction patchPrediction = farmingTracker.predictPatch(patch);
+		if (contract != null &&
+			patch.getRegion().getName().equalsIgnoreCase(FARMING_REGION_NAME) &&
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:131]: same as below; don't compare ui strings

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+	{
+		Widget npcDialog = client.getWidget(WidgetInfo.DIALOG_NPC_HEAD_MODEL);
+
+		if (npcDialog == null || npcDialog.getModelId() != GUILDMASTER_JANE_WIDGET_MODEL_ID)
+		{
+			return;
+		}
+
+		String dialogText = Text.removeTags(client.getWidget(WidgetInfo.DIALOG_NPC_TEXT).getText());
+		log.debug(dialogText + " ");
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:201]: ?

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+
+		if (dialogText.equals(CONTRACT_REWARDED))
+		{
+			setContract(null);
+			;
+		}
+		handleContractAssigning(dialogText);
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:208]: just remove these 4 lines, there is no reason for this to be its own method when you are just tail calling it

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+
+			if ((prediction.getProduce() != contract.getProduce() ||
+				(contract.requiresHealthCheck() && prediction.getCropState() == CropState.HARVESTABLE)) &&
+				emptyPatches == 0)
+			{
+				summary = SummaryState.OCCUPIED;
+				continue;
+			}
+
+			maxCompletionTime = Math.max(maxCompletionTime, prediction.getDoneEstimate());
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:302]: this is always `Math.max(0, v)`?

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+	private final ConfigManager configManager;
+
+	private FarmingContract contract = null;
+	private FarmingContractInfoBox infoBox;
+
+	@Getter
+	private long completionTime;
+
+	@Inject
+	public FarmingContractManager(
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:82]: you can just fieldinject all of these since your ctor is otherwise empty

File: runelite-client/src/main/java/net/runelite/client/util/Text.java (hunk 14 above)
Comment (melkypie) [reply runelite-client/src/main/java/net/runelite/client/util/Text.java:222]: I couldn't think of a better way to be able to access this for the infobox.

Comment (abextm) [reply runelite-client/src/main/java/net/runelite/client/util/Text.java:222]: just make it public where it used to be?

**Comment by melkypie** [comment]
Think that's all of them, changed the highlight to something more suitable 
image

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+import net.runelite.client.plugins.timetracking.Tab;
+import net.runelite.client.plugins.timetracking.TimeTrackingConfig;
+import net.runelite.client.plugins.timetracking.TimeTrackingPlugin;
+import net.runelite.client.ui.overlay.infobox.InfoBoxManager;
+import net.runelite.client.util.Text;
+
+@Slf4j
+public class FarmingContractManager
+{
+	private static final int GUILDMASTER_JANE_WIDGET_MODEL_ID = NullNpcID.NULL_8628;
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:53]:
```suggestion
	private static final int GUILDMASTER_JANE_NPC_ID = NullNpcID.NULL_8628;
```

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+
+		boolean hasEmptyPatch = false;
+		for (FarmingPatch patch : farmingWorld.getFarmingGuildRegion().getPatches())
+		{
+			if (patch.getImplementation() != patchImplementation)
+			{
+				continue;
+			}
+
+			PatchPrediction prediction = farmingTracker.predictPatch(patch);
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:243]: generally prefer having get + guards like these so the guard is attached to the get, as opposed to having a newline separating them:
```
			PatchPrediction prediction = farmingTracker.predictPatch(patch);
			if (prediction == null)
			{
				continue;
			}

			Produce produce = prediction.getProduce();
```

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+				{
+					summary = SummaryState.COMPLETED;
+					completionTime = 0;
+				}
+				else
+				{
+					summary = SummaryState.IN_PROGRESS;
+					completionTime = estimatedTime;
+				}
+				break;
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:285]: If you have 2 patches that match this crop, and the first one is growing, but the second one is done, this will say you have to wait, which seems wrong.

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java [hunk 24]
```diff
@@ -0,0 +1,322 @@
...
+		}
+		catch (NumberFormatException ignored)
+		{
+			return null;
+		}
+	}
+
+	private void setStoredContract(@Nullable Produce contract)
+	{
+
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java:306]: remove this newline

File: runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java (hunk 23 above)
Comment (melkypie) [reply runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingContractManager.java]: Tested this scenario (both ways) and some others after applying new fix.

**Comment by melkypie** [comment]
Changed some contract names but there are still some names that need to be tested.

**Comment by melkypie** [comment]
PR is fully ready for review. All of the dialog names have been checked either by me, Pharros or Hydrox

**Comment by abextm** [comment]
this should add the infobox if you start the plugin or login while in the farming guild

**Comment by melkypie** [comment]
It did work on login, but didn't work after plugin restarting due to the infobox not properly being deleted on shut down.

File: cache/src/main/java/net/runelite/cache/NpcManager.java [hunk 25]
```diff
@@ -97,14 +97,19 @@ public void java(File java) throws IOException
 		java.mkdirs();
 		try (IDClass ids = IDClass.create(java, "NpcID"))
 		{
-			for (NpcDefinition def : npcs.values())
+			try (IDClass nulls = IDClass.create(java, "NullNpcID"))
```
Comment (Adam-) [inline cache/src/main/java/net/runelite/cache/NpcManager.java:100]: For some reason the other classes don't do this, but you can have try-with-resources with multiple resources fyi.

**Comment by melkypie** [comment]
Changed other classes item/objects to use try-with multiple resources as was suggested with npc one.
Also put up an updated nullnpcid from abex
